File: db4oj/core/src/com/db4o/internal/cs/messages/MWriteNew.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o.internal.cs.messages;

import com.db4o.internal.*;
import com.db4o.internal.slots.*;

public final class MWriteNew extends MsgObject implements ServerSideMessage {
	
	public final boolean processAtServer() {
        int yapClassId = _payLoad.readInt();
        LocalObjectContainer stream = (LocalObjectContainer)stream();
        unmarshall(_payLoad._offset);
        synchronized (streamLock()) {
            ClassMetadata yc = yapClassId == 0 ? null : stream.classMetadataForId(yapClassId);
            
            int id = _payLoad.getID();
            stream.prefetchedIDConsumed(id);
            transaction().slotFreePointerOnRollback(id);
            
            Slot slot = stream.getSlot(_payLoad.length());
            _payLoad.address(slot.address());
            
            transaction().slotFreeOnRollback(id, slot);
            
            if(yc != null){
                yc.addFieldIndices(_payLoad,null);
            }
            stream.writeNew(yc, _payLoad);
            serverTransaction().writePointer( id, slot);
        }
        return true;
    }
}

File: db4on/core/src/core/com/db4o/internal/cs/messages/MWriteNew.cs
namespace com.db4o.@internal.cs.messages
{
	public sealed class MWriteNew : com.db4o.@internal.cs.messages.MsgObject
	{
		public sealed override bool ProcessAtServer(com.db4o.@internal.cs.ServerMessageDispatcher
			 serverThread)
		{
			int yapClassId = _payLoad.ReadInt();
			com.db4o.@internal.LocalObjectContainer stream = (com.db4o.@internal.LocalObjectContainer
				)Stream();
			Unmarshall(_payLoad._offset);
			lock (StreamLock())
			{
				com.db4o.@internal.ClassMetadata yc = yapClassId == 0 ? null : stream.GetYapClass
					(yapClassId);
				_payLoad.WriteEmbedded();
				int id = _payLoad.GetID();
				int length = _payLoad.GetLength();
				stream.PrefetchedIDConsumed(id);
				Transaction().SlotFreePointerOnRollback(id);
				int address = stream.GetSlot(length);
				_payLoad.Address(address);
				Transaction().SlotFreeOnRollback(id, address, length);
				if (yc != null)
				{
					yc.AddFieldIndices(_payLoad, null);
				}
				stream.WriteNew(yc, _payLoad);
				Transaction().WritePointer(id, address, length);
			}
			return true;
		}
	}
}
